syntax = "proto3";
package datasvc;
import "google/protobuf/empty.proto";

service DataProvider {
  rpc GetStatus(google.protobuf.Empty) returns (GetStatusReply) {}
  rpc ListData(google.protobuf.Empty) returns (ListDataReply) {}
  rpc UploadData(stream UploadRequest) returns (UploadReply) {}
  rpc BatchUpload(stream UploadRequest) returns (stream UploadReply) {}
  rpc DownloadData(DownloadRequest) returns (stream DownloadReply) {}
  rpc DeleteData(DownloadRequest) returns (UploadReply) {}

  rpc SendSharesData(SendSharesDataRequest) returns (SendSharesDataReply) {}
}

message DownloadRequest {
  string data_id = 1;
}

enum TaskStatus {
  Start = 0;
  Finished = 1;
  Cancelled = 2;
  Failed = 3;
}

message DownloadReply {
  oneof data {
    TaskStatus status = 1;
    bytes content = 2;
  }
}

message FileInfo {
  string file_name = 1;
  string file_type = 2;
  string description = 3;
  repeated string columns = 4;
  repeated string col_dtypes = 5;
  repeated string keywords = 6;
}

message UploadRequest {
  oneof data {
    FileInfo meta = 1;
    bytes content = 2;
  }
}

message UploadReply {
  bool ok = 1;
  string data_id = 2;
}

message ListDataReply {
  message Row {
    string data_id = 1;
    string info = 2;
    string file_name = 3;
    int32 size = 4;
  }
  repeated Row data = 1;
}

message GetDataDetailReply {
  message Item {
    string column = 1;
    string dtype = 2;
  }
  int32 n_rows = 1;
  repeated Item items = 2;
}

message GetStatusReply {
  string node_type = 1;
  string node_id = 2;
  string state = 3;
  message Info {
    string name = 1;
    string total_disk = 2;
    string used_disk = 3;
    string idle_disk = 4;
  }
  Info info = 4;
}

message SendSharesDataRequest {
  string data_id = 1;
  repeated string receivers = 2;
}

message SendSharesDataReply {
  TaskStatus status = 1;
}
